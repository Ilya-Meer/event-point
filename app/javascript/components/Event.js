import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import Card from 'react-bootstrap/Card';
import { addVote, removeVote } from '../utils/api';

const Event = ({ user, event }) => {
  const [voted, setVoted] = useState(false);
  const [numVotes, setNumVotes] = useState(event.votes.length);

  useEffect(() => {
    const voters = event.votes.map((v) => v.user_id);
    if (voters.includes(user.id)) {
      setVoted(true);
      setNumVotes(event.votes.length);
    }
  }, []);

  const handleVote = async () => {
    const payload = {
      user_id: user.id,
      event_id: event.id,
    };

    try {
      if (voted) {
        const res = await removeVote(payload);
        setNumVotes(numVotes - 1);
      } else {
        const res = await addVote(payload);
        setNumVotes(numVotes + 1);
      }
    } catch (error) {
      console.error(error);
    }

    setVoted(!voted);
  };

  return (
    <Card key={event.created_at} className='eventlist-event'>
      <Card.Body>
        <Card.Title>{event.topic}</Card.Title>
        <Card.Text>{event.description}</Card.Text>
        <Card.Text className='text-muted'>{`proposed by ${event.owner}`}</Card.Text>
        <Card.Text className='text-muted'>
          added on {event.created_at}
        </Card.Text>
        <div>
          <h3>
            Votes: <span>{numVotes}</span>
          </h3>
        </div>
        <div>
          <button onClick={handleVote}>{!voted ? 'Upvote' : 'Unvote'}</button>
        </div>
      </Card.Body>
    </Card>
  );
};

export default Event;

Event.propTypes = {
  event: PropTypes.shape({
    created_at: PropTypes.string,
    topic: PropTypes.string,
    description: PropTypes.string,
  }),
};
