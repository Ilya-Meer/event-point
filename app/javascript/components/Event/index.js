import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import Card from 'react-bootstrap/Card';
import Button from 'react-bootstrap/Button';
import { addVote, removeVote } from '../../utils/api';

const Event = ({ user, event, handleEditEvent, handleDeleteEvent }) => {
  const [voted, setVoted] = useState(false);
  const [numVotes, setNumVotes] = useState(event.votes.length);

  useEffect(() => {
    const voters = event.votes.map((v) => v.user_id);
    if (voters.includes(user.id)) {
      setVoted(true);
      setNumVotes(event.votes.length);
    }
  }, []);

  const handleVote = async () => {
    const payload = {
      user_id: user.id,
      event_id: event.id,
    };

    try {
      if (voted) {
        const res = await removeVote(payload);
        setNumVotes(numVotes - 1);
      } else {
        const res = await addVote(payload);
        setNumVotes(numVotes + 1);
      }
    } catch (error) {
      console.error(error);
    }

    setVoted(!voted);
  };

  return (
    <Card key={event.created_at} className='event'>
      <Card.Body>
        <div className='event-info'>
          <h2 className='event-info-title'>{event.topic}</h2>
          <p className='event-info-description text-muted'>
            {event.description}
          </p>
        </div>
        <div className='event-add-info'>
          <div className=''>
            <p className='event-add-info-heading text-muted'>Votes</p>
            <p data-testid='num-votes'>{numVotes}</p>
          </div>
          <div className=''>
            <p className='event-add-info-heading text-muted'>suggested by</p>
            <p>{event.owner}</p>
          </div>
        </div>
        <div className='event-button-container'>
          <Button
            onClick={handleVote}
            className={['upvote-button', `${voted ? 'upvoted' : ''}`].join(' ')}
          >
            {!voted ? 'Upvote' : 'Unvote'}
          </Button>
          {event.owner_id === user.id && (
            <div className='modify'>
              <Button
                variant='secondary'
                onClick={() => handleEditEvent(event.id)}
              >
                Edit
              </Button>
              <Button
                variant='danger'
                onClick={() => handleDeleteEvent(event.id)}
              >
                Delete
              </Button>
            </div>
          )}
        </div>
      </Card.Body>
    </Card>
  );
};

export default Event;

Event.propTypes = {
  user: PropTypes.object,
  event: PropTypes.shape({
    created_at: PropTypes.string,
    topic: PropTypes.string,
    description: PropTypes.string,
  }),
  handleEditEvent: PropTypes.func,
  handleDeleteEvent: PropTypes.func,
};
