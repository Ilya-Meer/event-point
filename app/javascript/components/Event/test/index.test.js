import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen, fireEvent, act } from '@testing-library/react';
import Event from '..';

it('renders without crashing', () => {
  const user = {};
  const event = {
    votes: [],
  };

  const tree = renderer.create(<Event user={user} event={event} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('displays vote information properly', () => {
  const user = {};
  const event = {
    votes: [
      {
        id: 1,
        user_id: 1,
        event_id: 1,
      },
    ],
  };

  render(<Event user={user} event={event} />);

  expect(screen.getByText(/Votes/)).toBeInTheDocument();
  expect(screen.getByText(/Votes/)).toHaveTextContent('Votes: 1');
});

it('increments vote information in local state if user has not voted', async () => {
  const user = { id: 5 };
  const event = {
    id: 1,
    votes: [
      {
        id: 1,
        user_id: 1,
        event_id: 1,
      },
    ],
  };

  fetchMock.once();

  render(<Event user={user} event={event} />);

  expect(screen.getByText(/Votes/)).toHaveTextContent('Votes: 1');

  await act(async () => {
    await fireEvent.click(screen.getByRole('button'));
  });

  expect(screen.getByText(/Votes/)).toHaveTextContent('Votes: 2');
  expect(fetchMock).toHaveBeenCalledTimes(1);

  const apiPayload = {
    body: '{"user_id":5,"event_id":1}',
    headers: { 'Content-Type': 'application/json' },
    method: 'POST',
  };

  expect(fetchMock).toHaveBeenCalledWith(
    'http://localhost:3000/add_vote',
    apiPayload
  );
});

it('decrements vote information in local state if user already voted', async () => {
  const user = { id: 1 };
  const event = {
    id: 1,
    votes: [
      {
        id: 1,
        user_id: 1,
        event_id: 1,
      },
    ],
  };

  fetchMock.once();

  render(<Event user={user} event={event} />);

  expect(screen.getByText(/Votes/)).toHaveTextContent('Votes: 1');

  await act(async () => {
    await fireEvent.click(screen.getByRole('button'));
  });

  expect(screen.getByText(/Votes/)).toHaveTextContent('Votes: 0');
  expect(fetchMock).toHaveBeenCalledTimes(1);

  const apiPayload = {
    body: '{"user_id":1,"event_id":1}',
    headers: { 'Content-Type': 'application/json' },
    method: 'POST',
  };

  expect(fetchMock).toHaveBeenCalledWith(
    'http://localhost:3000/remove_vote',
    apiPayload
  );
});
