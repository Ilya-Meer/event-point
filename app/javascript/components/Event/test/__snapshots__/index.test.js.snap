// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div
  className="event card"
>
  <div
    className="card-body"
  >
    <div
      className="event-info"
    >
      <h2
        className="event-info-title"
      />
      <p
        className="event-info-description text-muted"
      />
    </div>
    <div
      className="event-add-info"
    >
      <div
        className=""
      >
        <p
          className="event-add-info-heading text-muted"
        >
          Votes
        </p>
        <p
          data-testid="num-votes"
        >
          0
        </p>
      </div>
      <div
        className=""
      >
        <p
          className="event-add-info-heading text-muted"
        >
          suggested by
        </p>
        <p />
      </div>
    </div>
    <div
      className="event-button-container"
    >
      <button
        className="upvote-button  btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Upvote
      </button>
      <div
        className="modify"
      >
        <button
          className="btn btn-secondary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Edit
        </button>
        <button
          className="btn btn-danger"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Delete
        </button>
      </div>
    </div>
  </div>
</div>
`;
